vllm/compilation/backends.py

```diff
@@ -164,80 +166,212 @@ def wrap_inductor(graph, example_inputs, additional_inductor_config):
 
     if additional_inductor_config is not None:
         current_config.update(additional_inductor_config)
-    if current_config['post_grad_custom_post_pass'] is not None:
-        logger.warning(
-            "post_grad_custom_post_pass is already set in the config. "
-            "Overwriting it with the fix_functionalization")
-    current_config['post_grad_custom_post_pass'] = fix_functionalization
-    return compile_fx(graph, example_inputs, config_patches=current_config)
 
+    # inductor can inplace modify the graph, so we need to copy it
+    # see https://github.com/pytorch/pytorch/issues/138980
+    graph = copy.deepcopy(graph)
+    return compile_fx(graph, example_inputs, config_patches=current_config)
 
-def vllm_backend(
-        graph,
-        example_inputs,
-        additional_inductor_config: Optional[Dict] = None) -> Callable:
 
-    context = get_compile_context()
-    context = copy.deepcopy(context) if context is not None else []
-    sizes_to_specialize: List[int] = context
+def vllm_backend(graph, example_inputs) -> Callable:
```

Comment:
I think the "backend" is getting very complex. It might be worth separating it into a class, and maybe another class for `PiecewiseBackend`. Then all the callbacks could be methods and  the nonlocal variables can be members. Then keeping track of all of the state becomes much easier.

Reply:
good idea, added in [b366c2f](https://github.com/vllm-project/vllm/pull/9715/commits/b366c2ff4eabf272400f0c24fd117873c3654309)